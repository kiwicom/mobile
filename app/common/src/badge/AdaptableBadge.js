// @flow

import * as React from 'react';
import { View, Text, StyleSheet } from 'react-native';

import Color from '../Color';

type Props = {|
  text: string,
  color?: ?string,
|};

const style = StyleSheet.create({
  container: {
    alignSelf: 'flex-start',
    borderRadius: 2,
    paddingVertical: 2,
    paddingHorizontal: 5,
    backgroundColor: Color.grey.$500,
  },
  text: {
    color: '#fff',
  },
});

/**
 * This badge automatically adapt width based on the text length inside so the
 * result looks like this:
 *
 * .----------.
 * | VERIFIED |
 * `----------`
 *
 * And it expands with long text:
 *
 * .-------------------------------------------------.
 * | UNVERIFIED with very long text to make it clear |
 * `-------------------------------------------------`
 */
export default class AdaptableBadge extends React.Component<Props> {
  render = () => {
    const additionalContainerStyles = {};
    if (this.props.color) {
      additionalContainerStyles.backgroundColor = this.props.color;
    }

    return (
      <View style={[style.container, additionalContainerStyles]}>
        <Text style={style.text}>{this.props.text}</Text>
      </View>
    );
  };
}
