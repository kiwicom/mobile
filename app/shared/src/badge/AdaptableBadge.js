// @flow

import * as React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { type StylePropType } from '@kiwicom/react-native-app-shared';

import Color from '../Color';

type Props = {|
  text: string,
  style?: StylePropType,
  textStyle?: StylePropType,
|};

const style = StyleSheet.create({
  container: {
    alignSelf: 'flex-start',
    borderRadius: 2,
    paddingVertical: 2,
    paddingHorizontal: 5,
    marginRight: 5,
    marginBottom: 5,
    backgroundColor: Color.grey.$500,
  },
  text: {
    color: '#fff',
  },
});

/**
 * This badge automatically adapt width based on the text length inside so the
 * result looks like this:
 *
 * .----------.
 * | VERIFIED |
 * `----------`
 *
 * And it expands with long text:
 *
 * .-------------------------------------------------.
 * | UNVERIFIED with very long text to make it clear |
 * `-------------------------------------------------`
 */
export default function AdaptableBadge(props: Props) {
  return (
    <View style={[style.container, props.style]}>
      <Text style={[style.text, props.textStyle]}>{props.text}</Text>
    </View>
  );
}
