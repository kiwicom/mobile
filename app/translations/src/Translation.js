// @flow

import * as React from 'react';
import { NativeModules } from 'react-native';
import { Text } from '@kiwicom/react-native-app-shared';

import DefaultVocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      textTransform?: SupportedTransformationsType,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      textTransform?: SupportedTransformationsType,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  replaceValues = (text: string, values?: Object): string => {
    const VARIABLE_REGEX = /__([0-9]+_)?([a-zA-Z-_]+)__/g;
    return text.replace(
      VARIABLE_REGEX,
      (match, variableIndex, variableName) => {
        // 'variableIndex' is currently being ignored (we don't need it)
        const value = values && values[variableName];
        return value !== undefined ? value : '?';
      },
    );
  };

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const key = this.props.id;
    let translatedString = NativeModules.RNTranslationManager.translate(key);

    if (translatedString === key) {
      // fallback to our dummy vocabulary because native code didn't provide translation
      translatedString = DefaultVocabulary[key];
    }

    if (translatedString === undefined) {
      // Everything failed - native translation even our fallback. This usually
      // means we are using key that does not exist. In this case the only thing
      // we can actually do is to use original key as a text.
      translatedString = key;
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          this.replaceValues(translatedString, values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
