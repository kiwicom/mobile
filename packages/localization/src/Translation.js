// @flow

import * as React from 'react';
import { Text } from '@kiwicom/mobile-shared';
import { translate } from '@kiwicom/rnmodules';

import vocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';
import { getTranslation, replaceValues } from './TranslationHelpers';

// Preload translations
Object.keys(vocabulary).forEach(async key => {
  const nativeKey = 'mobile.' + key;

  const translatedString = await translate(nativeKey);
  if (translatedString != null) {
    vocabulary[key] = getTranslation(translatedString, key);
  }
});

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          replaceValues(vocabulary[this.props.id], values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
