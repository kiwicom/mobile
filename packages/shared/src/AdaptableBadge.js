// @flow strict

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/mobile-orbit';

import Text from './Text';
import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../types/Styles';
import type { TranslationType } from '../types/Translation';

type Props = {|
  +translation: TranslationType,
  +style?: StylePropType,
  +textStyle?: StylePropType,
  +icon?: React.Node,
  +type?: 'neutral' | 'info',
  +circled?: boolean,
|};

/**
 * This badge automatically adapt width based on the text length inside so the
 * result looks like this:
 *
 * .----------.
 * | VERIFIED |
 * `----------`
 *
 * And it expands with long text:
 *
 * .-------------------------------------------------.
 * | UNVERIFIED with very long text to make it clear |
 * `-------------------------------------------------`
 */
export default function AdaptableBadge(props: Props) {
  return (
    <View
      style={[
        styles.container,
        /* $FlowFixMe Errors after moving rn modules from untyped to
         * declarations */
        props.style,
        props.type === 'info' && styles.info,
        props.circled === true && styles.circled,
      ]}
    >
      {props.icon}
      <Text style={[styles.text, props.textStyle]}>{props.translation}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignSelf: 'flex-start',
    borderRadius: 2,
    paddingVertical: 2,
    paddingHorizontal: 5,
    backgroundColor: defaultTokens.paletteInkLight,
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  text: {
    color: defaultTokens.paletteWhite,
  },
  info: {
    backgroundColor: defaultTokens.paletteBlueNormal,
  },
  circled: {
    borderRadius: 11,
  },
});
