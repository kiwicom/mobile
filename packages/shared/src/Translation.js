// @flow

import * as React from 'react';
import {
  CaseTransform,
  getTranslation,
  type SupportedTransformationsType,
  type TranslationKeysType,
} from '@kiwicom/mobile-localization';

import Text from './Text';

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeysType,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of optional chaining
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */

// use this property to highlight translations for screenshoting or debugging
const highlightTranslations = false;
export default function Translation(props: Props) {
  const containerStyle = highlightTranslations
    ? {
        backgroundColor: props.id ? 'orange' : 'pink',
      }
    : undefined;
  if (props.id === undefined && props.passThrough != null) {
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(String(props.passThrough), props.textTransform)}
      </Text>
    );
  } else if (props.id === undefined) {
    return null;
  }

  const values = props.values || {};
  return (
    <Text style={containerStyle}>
      {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
      {CaseTransform(getTranslation(props.id, values), props.textTransform)}
    </Text>
  );
}
