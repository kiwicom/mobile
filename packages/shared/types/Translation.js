// @flow

import * as React from 'react';

import Translation from '../src/Translation';
import TranslationFragment from '../src/TranslationFragment';

export type TranslationType =
  | React.Element<typeof Translation>
  | React.Element<typeof TranslationFragment>;
