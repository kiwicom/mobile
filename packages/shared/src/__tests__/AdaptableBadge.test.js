// @flow

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/mobile-playground';

import AdaptableBadge from '../AdaptableBadge';
import Translation from '../Translation';

it('renders text', () => {
  PlaygroundRenderer.render(
    <AdaptableBadge translation={<Translation passThrough="default badge" />} />,
  );
});

it('renders text with color', () => {
  PlaygroundRenderer.render(
    <AdaptableBadge
      translation={<Translation passThrough="badge with color" />}
      style={{
        backgroundColor: 'red',
      }}
    />,
  );
});

it('renders very long texts', () => {
  PlaygroundRenderer.render(
    <AdaptableBadge
      translation={
        <Translation passThrough="very very very very very long badge (adaptable width)" />
      }
    />,
  );
});

it('renders correctly with type info', () => {
  PlaygroundRenderer.render(
    <AdaptableBadge translation={<Translation passThrough="type info" />} type="info" />,
  );
});

it('renders correctly with circled true', () => {
  PlaygroundRenderer.render(
    <AdaptableBadge translation={<Translation passThrough="circled" />} circled={true} />,
  );
});
