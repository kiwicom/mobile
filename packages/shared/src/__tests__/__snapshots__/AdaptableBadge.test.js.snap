// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders text 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "#7f91a8",
        "borderRadius": 2,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "paddingHorizontal": 5,
        "paddingVertical": 2,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
        },
        undefined,
      ]
    }
  >
    <Translation
      passThrough="default badge"
    />
  </Text>
</View>
`;

exports[`renders text with color 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "#7f91a8",
        "borderRadius": 2,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "paddingHorizontal": 5,
        "paddingVertical": 2,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
        },
        undefined,
      ]
    }
  >
    <Translation
      passThrough="badge with color"
    />
  </Text>
</View>
`;

exports[`renders very long texts 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "#7f91a8",
        "borderRadius": 2,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "paddingHorizontal": 5,
        "paddingVertical": 2,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
        },
        undefined,
      ]
    }
  >
    <Translation
      passThrough="very very very very very long badge (adaptable width)"
    />
  </Text>
</View>
`;
